From 977c5c6997261556ef657340f6634a966fa94199 Mon Sep 17 00:00:00 2001
From: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Date: Mon, 30 Oct 2017 18:00:06 +0200
Subject: [PATCH 140/767] staging: fsl-dpaa2/eth: Use Rx FD for XDP_TX

Discard the const qualifier of the ingress FD structure and
reuse it on XDP_TX, instead of building a new FD.

Signed-off-by: Ioana Radulescu <ruxandra.radulescu@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.09.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c | 17 +++++++----------
 1 file changed, 7 insertions(+), 10 deletions(-)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
index a2fd1fb336d3..39f927b13379 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
@@ -233,24 +233,20 @@ static struct sk_buff *build_frag_skb(struct dpaa2_eth_priv *priv,
 }
 
 static int dpaa2_eth_xdp_tx(struct dpaa2_eth_priv *priv,
-			    const struct dpaa2_fd *fd,
+			    struct dpaa2_fd *fd,
 			    void *buf_start,
 			    u16 queue_id)
 {
 	struct dpaa2_eth_fq *fq;
 	struct rtnl_link_stats64 *percpu_stats;
 	struct dpaa2_eth_drv_stats *percpu_extras;
-	struct dpaa2_fd tx_fd;
 	struct dpaa2_faead *faead;
 	u32 ctrl, frc;
 	int i, err;
 
-	/* Clone the FD, as we need to make changes to it */
-	memcpy(&tx_fd, fd, sizeof(tx_fd));
-
 	/* Mark the egress frame annotation area as valid */
-	frc = dpaa2_fd_get_frc(&tx_fd);
-	dpaa2_fd_set_frc(&tx_fd, frc | DPAA2_FD_FRC_FAEADV);
+	frc = dpaa2_fd_get_frc(fd);
+	dpaa2_fd_set_frc(fd, frc | DPAA2_FD_FRC_FAEADV);
 
 	ctrl = DPAA2_FAEAD_A4V | DPAA2_FAEAD_A2V | DPAA2_FAEAD_EBDDV;
 	faead = dpaa2_get_faead(buf_start);
@@ -263,7 +259,7 @@ static int dpaa2_eth_xdp_tx(struct dpaa2_eth_priv *priv,
 	fq = &priv->fq[queue_id];
 	for (i = 0; i < DPAA2_ETH_ENQUEUE_RETRIES; i++) {
 		err = dpaa2_io_service_enqueue_qd(NULL, priv->tx_qdid, 0,
-						  fq->tx_qdbin, &tx_fd);
+						  fq->tx_qdbin, fd);
 		if (err != -EBUSY)
 			break;
 	}
@@ -300,7 +296,7 @@ static void release_fd_buf(struct dpaa2_eth_priv *priv,
 
 static u32 dpaa2_eth_run_xdp(struct dpaa2_eth_priv *priv,
 			     struct dpaa2_eth_channel *ch,
-			     const struct dpaa2_fd *fd,
+			     struct dpaa2_fd *fd,
 			     u16 queue_id,
 			     void *vaddr)
 {
@@ -400,7 +396,8 @@ static void dpaa2_eth_rx(struct dpaa2_eth_priv *priv,
 	percpu_extras = this_cpu_ptr(priv->percpu_extras);
 
 	if (fd_format == dpaa2_fd_single) {
-		xdp_act = dpaa2_eth_run_xdp(priv, ch, fd, queue_id, vaddr);
+		xdp_act = dpaa2_eth_run_xdp(priv, ch, (struct dpaa2_fd *)fd,
+					    queue_id, vaddr);
 		if (xdp_act != XDP_PASS)
 			return;
 
-- 
2.17.0

